pkg/installation/install.go

```diff
@@ -71,42 +66,79 @@ func Install(p environment.Paths, plugin index.Plugin, forceDownloadFile string)
 		return errors.Wrap(err, "failed to look up plugin receipt")
 	}
 
-	glog.V(1).Infof("Finding download target for plugin %s", plugin.Name)
-	version, sha256, uri, fos, bin, err := getDownloadTarget(plugin)
+	// Find available installation candidate
+	candidate, ok, err := GetMatchingPlatform(plugin.Spec.Platforms)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed trying to find a matching platform in plugin spec")
+	}
+	if !ok {
+		return errors.Wrapf(err, "plugin %q does not offer installation for this platform", plugin.Name)
 	}
 
 	// The actual install should be the last action so that a failure during receipt
 	// saving does not result in an installed plugin without receipt.
-	glog.V(3).Infof("Install plugin %s", plugin.Name)
-	if err := install(plugin.Name, version, sha256, uri, bin, p, fos, forceDownloadFile); err != nil {
+	glog.V(3).Infof("Install plugin %s at version=%s", plugin.Name, plugin.Spec.Version)
+	if err := install(installOperation{
+		pluginName: plugin.Name,
+		platform:   candidate,
+
+		downloadStagingDir: filepath.Join(p.DownloadPath(), plugin.Name),
+		binDir:             p.BinPath(),
+		installDir:         p.PluginVersionInstallPath(plugin.Name, plugin.Spec.Version),
+	}, opts); err != nil {
 		return errors.Wrap(err, "install failed")
 	}
 	glog.V(3).Infof("Storing install receipt for plugin %s", plugin.Name)
 	err = receipt.Store(plugin, p.PluginInstallReceiptPath(plugin.Name))
 	return errors.Wrap(err, "installation receipt could not be stored, uninstall may fail")
 }
 
-func install(plugin, version, sha256sum, uri, bin string, p environment.Paths, fos []index.FileOperation, forceDownloadFile string) error {
-	dst, err := downloadAndMove(version, sha256sum, uri, fos, filepath.Join(p.DownloadPath(), plugin), p.PluginInstallPath(plugin), forceDownloadFile)
-	if err != nil {
-		return errors.Wrap(err, "failed to download and move during installation")
+func install(op installOperation, opts InstallOpts) error {
+	// Download and extract
+	glog.V(3).Infof("Creating download staging directory %q", op.downloadStagingDir)
+	if err := os.MkdirAll(op.downloadStagingDir, 0755); err != nil {
+		return errors.Wrapf(err, "could not create download path %q", op.downloadStagingDir)
+	}
+	defer func() {
+		glog.V(3).Infof("Deleting the download staging directory %s", op.downloadStagingDir)
+		_ = os.RemoveAll(op.downloadStagingDir)
+	}()
+	if err := downloadAndExtract(op.downloadStagingDir, op.platform.URI, op.platform.Sha256, opts.ArchiveFileOverride); err != nil {
+		return errors.Wrap(err, "failed to download and extract")
 	}
 
-	subPathAbs, err := filepath.Abs(dst)
+	if err := moveToInstallDir(op.downloadStagingDir, op.installDir, op.platform.Files); err != nil {
+		return errors.Wrap(err, "failed while moving files to the installation directory")
+	}
+
+	subPathAbs, err := filepath.Abs(op.installDir)
```

Comment:
Could you extract the whole sub-path checking logic to a separate function?

Reply:
Can we do that in another PR? 😄

Reply:
sure :)